gcputil/credentials.go

```diff
@@ -50,6 +53,127 @@ type GcpCredentials struct {
 	ProjectId    string `json:"project_id" structs:"project_id" mapstructure:"project_id"`
 }
 
+type ExternalAccountCredential struct {
+	// External Account fields
+	Audience              string   `json:"audience"`
+	ServiceAccountEmail   string   `json:"service_account_email"`
+	SubjectTokenType      string   `json:"subject_token_type"`
+	TokenURL              string   `json:"token_url"`
+	Scopes                []string `json:"scopes"`
+	WorkloadIdentityToken string   `json:"workload_identity_token"`
+}
+
+func (c *ExternalAccountCredential) TokenSource(ctx context.Context) (oauth2.TokenSource, error) {
+	ts := tokenSource{
+		ctx:    ctx,
+		config: c,
+	}
+	return oauth2.ReuseTokenSource(nil, ts), nil
+}
+
+func (c *ExternalAccountCredential) GetCredentials(ctx context.Context) (*google.Credentials, error) {
+	ts, err := c.TokenSource(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &google.Credentials{
+		TokenSource: ts,
+	}, nil
+}
+
+// tokenSource is the source that handles external credentials. It is used to retrieve Tokens.
+type tokenSource struct {
+	ctx    context.Context
```

Comment:
In general, it isn't recommended to store the context in a struct since it can make the make the scope and lifetime unclear. Instead, we can pass the context to each function explicitly. See https://go.dev/blog/context-and-structs